cmd/goal/application.go

```diff
@@ -232,134 +273,175 @@ func newBoxRef(arg string) boxRef {
-			// transaction as the full number. Though it's hard to see why.
-			if !found && tbr.appID == uint64(appIdx) {
-				index = 0
-				found = true
-			}
-			if !found {
-				reportErrorf("Box ref with appId (%d) not in foreign-apps", tbr.appID)
-			}
+	if both {
+		appID := parseUInt64(one, "app id", "local ref")
+		return localRef{
+			appID:   appID,
+			address: two,
 		}
-		output[i] = transactions.BoxRef{
-			Index: index,
-			Name:  rawName,
+	}
+
+	// one is missing, so we should have a number or an address.  Try to parse
+	// it as a number. If it fails, assume an address, because at this stage we
+	// don't parse addresses.
+	if appID, err := strconv.ParseUint(one, 10, 64); err == nil {
+		return localRef{
+			appID:   appID,
+			address: "",
 		}
 	}
-	return output
+
+	return localRef{
+		appID:   0,
+		address: one,
+	}
 }
 
-func parseAppInputs(inputs appCallInputs) (args [][]byte, accounts []string, foreignApps []uint64, foreignAssets []uint64, boxes []transactions.BoxRef) {
-	accounts = inputs.Accounts
-	foreignApps = inputs.ForeignApps
-	foreignAssets = inputs.ForeignAssets
-	boxes = translateBoxRefs(inputs.Boxes, foreignApps)
-	args = make([][]byte, len(inputs.Args))
-	for i, arg := range inputs.Args {
+// parseAppInputs converts inputs from a very textual input form (coming from
+// CLI or a JSON file), to a more strongly typed form, using the various "real"
+// types from `basics`.
+func parseAppInputs(inputs appCallInputs) ([][]byte, libgoal.RefBundle) {
+	args := make([][]byte, 0, len(inputs.Args))
+	for _, arg := range inputs.Args {
 		rawValue, err := arg.Raw()
 		if err != nil {
-			reportErrorf("Could not decode input at index %d: %v", i, err)
+			reportErrorf("Could not decode app-arg %s:%s: %v", arg.Encoding, arg.Value, err)
 		}
-		args[i] = rawValue
+		args = append(args, rawValue)
 	}
-	return
+	locals := util.Map(inputs.Locals, func(hr localRef) basics.LocalRef {
+		return basics.LocalRef{
+			App:     basics.AppIndex(hr.appID),
+			Address: cliAddress(hr.address)}
+	})
+	holdings := util.Map(inputs.Holdings, func(hr holdingRef) basics.HoldingRef {
+		return basics.HoldingRef{
+			Asset:   basics.AssetIndex(hr.assetID),
+			Address: cliAddress(hr.address)}
+	})
+	boxes := util.Map(inputs.Boxes, func(br boxRef) basics.BoxRef {
+		rawName, err := br.name.Raw()
+		if err != nil {
+			reportErrorf("Could not decode box name %s: %v", br.name, err)
+		}
+		return basics.BoxRef{App: basics.AppIndex(br.appID), Name: string(rawName)}
+	})
+	refs := libgoal.RefBundle{
+		UseAccess: inputs.UseAccess,
+		Accounts:  util.Map(inputs.Accounts, cliAddress),
+		Apps:      util.Map(inputs.ForeignApps, func(idx uint64) basics.AppIndex { return basics.AppIndex(idx) }),
+		Assets:    util.Map(inputs.ForeignAssets, func(idx uint64) basics.AssetIndex { return basics.AssetIndex(idx) }),
+
+		Locals:   locals,
+		Holdings: holdings,
+		Boxes:    boxes,
+	}
+	return args, refs
+}
+
+func cliAddress(acct string) basics.Address {
```

Comment:
Probably worthy of a simple unit test given frequency of usage.

Reply:
Fair enough. app(xxx) does get used in e2e tests, but that is a nice enough function that I can easily add unit tests.